activerecord/lib/active_record/persistence.rb

```diff
@@ -181,7 +181,11 @@ def destroy
       _raise_readonly_record_error if readonly?
       destroy_associations
       self.class.connection.add_transaction_record(self)
-      destroy_row if persisted?
+      @_actually_destroyed = if persisted?
```

Comment:
The issue is somehow flagging whether the `SQL DELETE` actually affected any rows (so we can trigger the callback then and only then). To do so we need to store the result of `destroy_row` somewhere. Right now we have:

```
    def persisted?
      sync_with_transaction_state
      !(@new_record || @destroyed)
    end
```

So by De Morgan's law your suggestion ends up being:

```
if persisted?
   @destroyed = destroy_row > 0
end
```

The problem is that does not work - it will fail the following test: https://github.com/rails/rails/blob/master/activerecord/test/cases/associations/bidirectional_destroy_dependencies_test.rb#L34-L39